### DISCUSSION TITLE
Rated torque

### DISCUSSION THREAD

**Discussion by OxygenLiu** [dzid26/StepperServoCAN#20] [Ideas]
The rate current of my NEMA17 stepper motor is 1.3A, and rated torque 0.45Nm. It is assumed a linear torque/amp ratio of 0.45Nm/1.3A. With 3.3A peak phase current, the expected torque is about 1.14Nm. 

I am wondering if the max phase current 3.3A will satuate the permanent magnet, leading to a much lower torque than expected. It would be ideal to have a dynometer like [this](https://build-its.blogspot.com/2019/12/motor-dynamometer.html) to evaluate the controller and motor performance in detail.

**Comment by dzid26** [comment]
Permanent magnet, probably not. 
The steel core, possibly. But from my experience, these motors do produce more torque with higher currents, though the increase may not be exactly linear as you are suggesting.  Another limitation is thermal at this kind of current - short term for the PCB and quickly after for the motor. Additionally hot motor has higher copper resistance so it naturally starts to limit max torque. The value could be estimated, but in practice, we have close-loop torque control that takes care of those small differences.  

If you need more torque, you may have to either use a motor with a bit more windings (but less than 4ohm per phase) or a bigger motor.  I found some Nema17 stepper of the same length with 7.5ohm ([link](https://www.ebay.com/itm/255428289790), [link](https://www.lamtechnologies.com/Product.aspx?lng=EN&idp=m1173030) [link](https://www.sensorgruppen.se/wp-content/uploads/2018/11/M11730xx_DSUK.pdf)) windings (which is probably too much for 12V, so stall torque will probably be less than for your motor, but much cooler 14V/7 = 2amps max), and I am going to test what are limitations of going too many winding. For one, with this amount of winding and inductance max speed will be lower.  If that "extreme" test is somewhat successful, that means that another 4ohm motor [LDO-42STH40-1004ASC](https://lecktor.com/en/stepper-motors/1310-18-nema17-stepper-motor-ldo-42sth40-1004asc.html) (of the same size so it should be possible to simply swap the stator) should also work fine  - and this could be an ultimate motor for 14.4V that is in the car. 


The dynamometer would be the best to characterize the motor precisely.  For stall torque really just a scale (load cell) is needed at the known length of the arm. 
For spinning torque, I was also thinking of a simpler eddy current copper disk brake device with an electromagnet varying the load and a load cell. It would provide a very smooth load but perhaps would be useless at low speed.  Everything takes time though. :)

**Comment by dzid26** [comment]
OK, I quickly tried the 7.5ohm/15mH motor. No surprise, it has good torque at a low current. But top stall torque is limited by the max current possible (ohms law). - Additionally, torque ripple is very bad at high current, since the control is not trying to predict this limitation - so two phases produce more current than a single one.
The no-load speed is lower, but still is very high: 75rev/s (vs 125rev/s on ~5mH motor), so no issue there. 

So I would say go ahead and try swapping your stator with one from the [ldo-42sth40-1004asc](https://www.printedsolid.com/products/ldo-nema-17-motor-cool-power-qlink-ldo-42sth40-1004asc). It has 4ohms@11mH windings. Higher resistance and inductance indicate there are physically more winding turns in this stator, which results in higher torque for the same current. Since this motor is also 40mm long, the stator is the direct replacement of the default geared motor. In theory, it has about **38% more stall torque** in the same form factor. It is very close to optimal in terms of torque density for 14.4V. 
![image](https://github.com/dzid26/StepperServoCAN/assets/841061/9fef649f-d3b6-48e1-94e3-ae9dccc5c70e)

One thing that may suffer with a high impedance motor like this, is peak power - meaning **under load** maximum rotation speed can be lower. However, for steering, it doesn't matter so much, as the quickest rotation is required when straightening the wheel (low load). 

I am curious if this stator would let you rotate a complete turn in your car at idle.

**Reply by dzid26** [reply]
Currently, I am getting maybe 10% more torque with [ldo-42sth40-1004asc](https://www.printedsolid.com/products/ldo-nema-17-motor-cool-power-qlink-ldo-42sth40-1004asc) stator.  The number of winding turns is greater (there is more copper wire).  But it might be that we are reaching magnetic saturation limits of Nema 17 rotors. It's hard to precisely measure max torque without #21 being resolved.

**Reply by dzid26** [reply]
Torque curve for the ldo-42sth40-1004asc at 14V (from what I remember). [Motor torque LDO.ods](https://github.com/dzid26/StepperServoCAN/files/14760519/Motor.torque.LDO.ods)

**Comment by dzid26** [comment]
OxygenLiu, I converted the issue into a discussion and changed the title. 
Let me paste here my post from our private conversation, for others to see:

> Recently, I characterized the motor on a rig with a scale. Well, the motor had a different rotor, no gearbox, and only the stator was from 42BYG 17HS4401S.
> It shouldn't make a difference though, and it was much easier to measure - without gearbox multiplication.
> 
> It seems that the actual stall torque is about half of the rated one.  Not that it matters too much for the control (due to close loop and feedforward calibration in Openpilot), but it means not trusting rated torque too much.
> 
> ![image](https://github.com/dzid26/StepperServoCAN/assets/841061/6d905311-83f5-4fa0-ba43-cd478f722529)
> [Motor torque BYG.ods](https://github.com/dzid26/StepperServoCAN/files/14756739/Motor.torque.BYG.ods)

> 
For now, I am still using rated torque (and not actual torque form measurements) when talking about motor torque. 
I will have to repeat this measurement in the future, to make sure I didn't make a mistake.

**Comment by OxygenLiu** [comment]
I am wondering if you measured the actual current per phase?

**Reply by dzid26** [reply]
I did, but I failed to include it in the analysis above since I am currently measuring externally.
- It does account for some of the discrepancy. 
2Amps commanded to A4950 (1.995V read on Vref pin) corresponds to 1.78A measured. 

Using experimental voltage-control (I = Vin*duty% / R) instead of chip current limit, the difference is even bigger - but that is to be expected as I am not compensating for blanking time, etc. 

I will want to connect a load-cell to the mcu, enable onboard ADC [current measurement](https://github.com/dzid26/StepperServo-hardware/issues/2#issuecomment-1582306255), and have a test case that will do a current sweep and record everything.

**Comment by OxygenLiu** [comment]
It would be interesting to see the results of `LSS_A_ADC` and `LSS_B_ADC`.  When 10xV<sub>S</sub> > V<sub>REF</sub>, the mixed decay mode is activated, which may be the culpt of unwanted high current behavior.
